exchanges/ftx/ftx_websocket.go

```diff
@@ -0,0 +1,485 @@
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/ticker"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/websocket/wshandler"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/websocket/wsorderbook"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const (
+	ftxWSURL          = "wss://ftx.com/ws/"
+	ftxWebsocketTimer = 13 * time.Second
+	wsTicker          = "ticker"
+	wsTrades          = "trades"
+	wsOrderbook       = "orderbook"
+	wsMarkets         = "markets"
+	wsFills           = "fills"
+	wsOrders          = "orders"
+	wsUpdate          = "update"
+	wsPartial         = "partial"
+	subscribe         = "subscribe"
+	unsubscribe       = "unsubscribe"
+)
+
+// WsConnect connects to a websocket feed
+func (f *FTX) WsConnect() error {
+	if !f.Websocket.IsEnabled() || !f.IsEnabled() {
+		return errors.New(wshandler.WebsocketNotEnabled)
+	}
+	var dialer websocket.Dialer
+	err := f.WebsocketConn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	f.WebsocketConn.SetupPingHandler(wshandler.WebsocketPingHandler{
+		MessageType: websocket.PingMessage,
+		Delay:       ftxWebsocketTimer,
+	})
+	if f.Verbose {
+		log.Debugf(log.ExchangeSys, "%s Connected to Websocket.\n", f.Name)
+	}
+	go f.wsReadData()
+	if f.GetAuthenticatedAPISupport(exchange.WebsocketAuthentication) {
+		err := f.WsAuth()
+		if err != nil {
+			f.Websocket.DataHandler <- err
+			f.Websocket.SetCanUseAuthenticatedEndpoints(false)
+		}
+	}
+	f.GenerateDefaultSubscriptions()
+	return nil
+}
+
+// WsAuth sends an authentication message to receive auth data
+func (f *FTX) WsAuth() error {
+	intNonce := time.Now().UnixNano() / 1000000
+	strNonce := strconv.FormatInt(intNonce, 10)
+	hmac := crypto.GetHMAC(
+		crypto.HashSHA256,
+		[]byte(strNonce+"websocket_login"),
+		[]byte(f.API.Credentials.Secret),
+	)
+	sign := crypto.HexEncodeToString(hmac)
+	req := Authenticate{Operation: "login",
+		Args: AuthenticationData{
+			Key:  f.API.Credentials.Key,
+			Sign: sign,
+			Time: intNonce,
+		},
+	}
+	return f.WebsocketConn.SendJSONMessage(req)
+}
+
+// Subscribe sends a websocket message to receive data from the channel
+func (f *FTX) Subscribe(channelToSubscribe wshandler.WebsocketChannelSubscription) error {
+	var sub WsSub
+	a, err := f.GetPairAssetType(channelToSubscribe.Currency)
+	if err != nil {
+		return err
+	}
+	switch channelToSubscribe.Channel {
+	case wsFills, wsOrders:
+		sub.Operation = subscribe
+		sub.Channel = channelToSubscribe.Channel
+	default:
+		sub.Operation = subscribe
+		sub.Channel = channelToSubscribe.Channel
+		sub.Market = f.FormatExchangeCurrency(channelToSubscribe.Currency, a).String()
+	}
+	return f.WebsocketConn.SendJSONMessage(sub)
+}
+
+// GenerateDefaultSubscriptions generates default subscription
+func (f *FTX) GenerateDefaultSubscriptions() {
+	var channels = []string{wsTicker, wsTrades, wsOrderbook, wsMarkets, wsFills, wsOrders}
```

Comment:
fills and orders need to be separated and checked if websocket has authentication enabled or these will send and error like so: INCOMING: {"type": "error", "code": 400, "msg": "Not logged in"}
You are also sending multiple subscriptions:
```
SUBS: {"channel":"ticker","market":"BTC/USD","op":"subscribe"}
SUBS: {"channel":"trades","market":"BTC/USD","op":"subscribe"}
SUBS: {"channel":"orderbook","market":"BTC/USD","op":"subscribe"}
SUBS: {"channel":"markets","market":"BTC/USD","op":"subscribe"}
SUBS: {"channel":"fills","op":"subscribe"}
SUBS: {"channel":"orders","op":"subscribe"}
SUBS: {"channel":"ticker","market":"THETA-PERP","op":"subscribe"}
SUBS: {"channel":"trades","market":"THETA-PERP","op":"subscribe"}
SUBS: {"channel":"orderbook","market":"THETA-PERP","op":"subscribe"}
SUBS: {"channel":"markets","market":"THETA-PERP","op":"subscribe"}
SUBS: {"channel":"fills","op":"subscribe"}
SUBS: {"channel":"orders","op":"subscribe"}
```
Also channel Markets does **NOT** need an associative pair, as contracts expire a partial list will be sent and should be handled appropriately.